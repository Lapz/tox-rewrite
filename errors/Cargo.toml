[package]
name = "errors"
version = "0.1.0"
authors = ["Lenard Pratt <l3np27@gmail.com>"]
edition = "2018"

# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html

[dependencies]
codespan-reporting = {version="^0.9.0",git="https://github.com/Lapz/codespan"}
reporting = {version="^0.9.0",package="codespan-reporting"}
itertools="*"
salsa = "^0.14.1"