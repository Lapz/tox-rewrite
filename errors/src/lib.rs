pub mod db;
mod files;
pub mod pos;
mod reporter;
pub use crate::reporter::Reporter;
pub use codespan_reporting::{
    diagnostic::Diagnostic,
    files::Files,
    term::{
        emit,
        termcolor::{ColorChoice, StandardStream},
        Config,
    },
};

pub use pos::Span;

pub use db::{FileDatabase, FileDatabaseStorage, FileId};
pub type WithError<T> = Result<T, Vec<Diagnostic<FileId>>>;
