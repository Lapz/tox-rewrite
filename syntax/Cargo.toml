[package]
name = "syntax"
version = "0.1.0"
authors = ["Lenard Pratt <l3np27@gmail.com>"]
edition = "2018"

# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html

[dependencies]
rowan = "0.9.0"
itertools="*"
errors = {path="../errors"}

[dev-dependencies]
insta = "0.8.2"
tempfile = "3.1.0"