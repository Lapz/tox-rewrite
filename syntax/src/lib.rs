#[macro_use]
mod macros;

pub mod ast;
mod lexer;

mod ast_ext;
mod token;
mod traits;
pub use ast::SyntaxKind;
pub use lexer::Lexer;
pub use rowan::{SmolStr, TextRange, TextUnit};
pub use token::Token;
pub use traits::*;

pub type SyntaxNode = rowan::SyntaxNode<ToxLang>;
#[allow(unused)]
pub type SyntaxToken = rowan::SyntaxToken<ToxLang>;
#[allow(unused)]
pub type SyntaxElement = rowan::NodeOrToken<SyntaxNode, SyntaxToken>;

pub type SyntaxNodeChildren = rowan::SyntaxNodeChildren<ToxLang>;
#[derive(Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord, Hash)]
pub enum ToxLang {}
impl rowan::Language for ToxLang {
    type Kind = SyntaxKind;
    fn kind_from_raw(raw: rowan::SyntaxKind) -> Self::Kind {
        assert!(raw.0 <= SyntaxKind::__LAST as u16);
        unsafe { std::mem::transmute::<u16, SyntaxKind>(raw.0) }
    }
    fn kind_to_raw(kind: Self::Kind) -> rowan::SyntaxKind {
        kind.into()
    }
}
