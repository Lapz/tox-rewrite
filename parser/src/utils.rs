use crate::AstNode;

#[cfg(test)]
use syntax::ast::SourceFile;

pub fn dump_debug<T: AstNode>(item: &T) -> String {
    format!("{:#?}", item.syntax())
}

#[cfg(test)]
#[salsa::database(errors::FileDatabaseStorage, crate::ParseDatabaseStorage)]
#[derive(Debug, Default)]
pub struct MockDatabaseImpl {
    runtime: salsa::Runtime<MockDatabaseImpl>,
}

#[cfg(test)]
impl salsa::Database for MockDatabaseImpl {
    fn salsa_runtime(&self) -> &salsa::Runtime<MockDatabaseImpl> {
        &self.runtime
    }

    fn salsa_runtime_mut(&mut self) -> &mut salsa::Runtime<MockDatabaseImpl> {
        &mut self.runtime
    }
}

#[cfg(test)]
pub fn parse<'a>(input: &'a str) -> SourceFile {
    use crate::ParseDatabase;
    use errors::FileDatabase;
    use std::io::Write;
    use tempfile::NamedTempFile;

    let mut file = NamedTempFile::new().unwrap();
    write!(file, "{}", input).unwrap();
    let db = MockDatabaseImpl::default();
    let handle = db.intern_file(file.path().to_path_buf());

    db.parse(handle).unwrap()
}
