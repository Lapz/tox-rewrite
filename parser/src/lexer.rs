#[cfg(test)]
mod tests {

    use crate::utils::MockDatabaseImpl;
    use crate::ParseDatabase;
    use errors::{pos::Span, FileDatabase};
    use insta::assert_debug_snapshot;
    use std::io::Write;
    use syntax::Token;
    use tempfile::NamedTempFile;

    fn get_tokens(input: &str) -> Vec<Span<Token>> {
        let mut file = NamedTempFile::new().unwrap();
        write!(file, "{}", input).unwrap();

        let db = MockDatabaseImpl::default();
        let handle = db.intern_file(file.path().to_path_buf());

        db.lex(handle).unwrap()
    }

    #[test]
    fn it_works() {
        let input = "fn main() {print(\"hello\")}";

        assert_debug_snapshot!(get_tokens(input))
    }
    #[test]
    fn lex_int() {
        let input = "12,34,,45,67,89,0";

        assert_debug_snapshot!(get_tokens(input))
    }

    #[test]
    fn lex_floats() {
        let input = "12.34,45.67,89.10";

        assert_debug_snapshot!(get_tokens(input))
    }

    #[test]
    fn lex_block_comments() {
        let input = "12 /*this is a comment */ 34";

        assert_debug_snapshot!(get_tokens(input))
    }

    #[test]
    fn lex_nested_block_comments() {
        let input = "/*this /*is a nested */  comment */";

        assert_debug_snapshot!(get_tokens(input))
    }

    #[test]
    fn lex_line_comment() {
        let input = "12 // this is a line comment";

        assert_debug_snapshot!(get_tokens(input))
    }
}
