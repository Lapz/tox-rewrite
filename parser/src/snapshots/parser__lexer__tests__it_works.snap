---
source: parser/src/lexer.rs
expression: get_tokens(input)
---
[
    Span {
        value: Token {
            kind: FN_KW,
            len: 2,
        },
        start: Position {
            column: 1,
            line: 1,
            absolute: 0,
        },
        end: Position {
            column: 3,
            line: 1,
            absolute: 2,
        },
    },
    Span {
        value: Token {
            kind: WHITESPACE,
            len: 1,
        },
        start: Position {
            column: 3,
            line: 1,
            absolute: 2,
        },
        end: Position {
            column: 4,
            line: 1,
            absolute: 3,
        },
    },
    Span {
        value: Token {
            kind: IDENT,
            len: 4,
        },
        start: Position {
            column: 4,
            line: 1,
            absolute: 3,
        },
        end: Position {
            column: 8,
            line: 1,
            absolute: 7,
        },
    },
    Span {
        value: Token {
            kind: L_PAREN,
            len: 1,
        },
        start: Position {
            column: 8,
            line: 1,
            absolute: 7,
        },
        end: Position {
            column: 8,
            line: 1,
            absolute: 7,
        },
    },
    Span {
        value: Token {
            kind: R_PAREN,
            len: 1,
        },
        start: Position {
            column: 9,
            line: 1,
            absolute: 8,
        },
        end: Position {
            column: 9,
            line: 1,
            absolute: 8,
        },
    },
    Span {
        value: Token {
            kind: WHITESPACE,
            len: 1,
        },
        start: Position {
            column: 10,
            line: 1,
            absolute: 9,
        },
        end: Position {
            column: 11,
            line: 1,
            absolute: 10,
        },
    },
    Span {
        value: Token {
            kind: L_CURLY,
            len: 1,
        },
        start: Position {
            column: 11,
            line: 1,
            absolute: 10,
        },
        end: Position {
            column: 11,
            line: 1,
            absolute: 10,
        },
    },
    Span {
        value: Token {
            kind: IDENT,
            len: 5,
        },
        start: Position {
            column: 12,
            line: 1,
            absolute: 11,
        },
        end: Position {
            column: 17,
            line: 1,
            absolute: 16,
        },
    },
    Span {
        value: Token {
            kind: L_PAREN,
            len: 1,
        },
        start: Position {
            column: 17,
            line: 1,
            absolute: 16,
        },
        end: Position {
            column: 17,
            line: 1,
            absolute: 16,
        },
    },
    Span {
        value: Token {
            kind: STRING,
            len: 7,
        },
        start: Position {
            column: 18,
            line: 1,
            absolute: 17,
        },
        end: Position {
            column: 25,
            line: 1,
            absolute: 24,
        },
    },
    Span {
        value: Token {
            kind: R_PAREN,
            len: 1,
        },
        start: Position {
            column: 25,
            line: 1,
            absolute: 24,
        },
        end: Position {
            column: 25,
            line: 1,
            absolute: 24,
        },
    },
    Span {
        value: Token {
            kind: R_CURLY,
            len: 1,
        },
        start: Position {
            column: 26,
            line: 1,
            absolute: 25,
        },
        end: Position {
            column: 26,
            line: 1,
            absolute: 25,
        },
    },
    Span {
        value: Token {
            kind: EOF,
            len: 1,
        },
        start: Position {
            column: 27,
            line: 1,
            absolute: 26,
        },
        end: Position {
            column: 27,
            line: 1,
            absolute: 26,
        },
    },
]
